linode/manager
# feat: [M3-7035] - Add DC-specific pricing to Linode backups

## Description 📝
Pricing for backups is now Region-dependent. This PR uses the mocked API response for the linode/types endpoint to get the correct price for a backup add-on with region_prices data. 

## Major Changes 🔄
With the `dcSpecificPricing` feature flag *on* and MSW *on*:
- In the Linode Create flow, adding Backups as an Add-On displays DC-specific backup pricing based on the selected Region in the Add-Ons panel and the checkout summary.
- On the Linode Details -> Backups tab:
   - The enable backups empty state copy displays DC-specific backup pricing based on the Linode's region.
   - On the Enable Backups confirmation dialog, copy displays correctly calculated DC-specific backup pricing based on the Linode's region.
- On the Enable All Backups drawer: 
   - Total cost of backups, and each Linode's DC-specific backup price, is displayed and calculated correctly.
   - Minor copy updates  (removed "Confirm to add backups to N Linode(s)" and prefixed the total with "Total for N Linodes:")
   - A Region column exists in the table with populated region data for each Linode
   - Modal is widened to fit all four columns (existing + new Regions column)
- Updated unit tests

## Preview 📷

| Flow/Component | Before  | After  |
| ------- | ------- | ------- |
| Linode Create | ![Screenshot 2023-08-28 at 11 46 43 AM](https://github.com/linode/manager/assets/114685994/8052744d-c8b8-4042-9d2a-b503fd01c597) | ![Screenshot 2023-08-28 at 9 49 39 AM](https://github.com/linode/manager/assets/114685994/69696a38-4ba9-459a-b362-98d9c18bd815) |
| Linode Details -> Backups | ![Screenshot 2023-08-28 at 11 45 33 AM](https://github.com/linode/manager/assets/114685994/05089635-9ec2-4dc2-bc41-8c5f26d461fc) | ![Screenshot 2023-08-28 at 9 51 25 AM](https://github.com/linode/manager/assets/114685994/f433ce28-d0d0-4174-a0e9-fdbb1e3e985b) |
| Enable Backups Dialog | ![Screenshot 2023-08-28 at 11 45 41 AM](https://github.com/linode/manager/assets/114685994/4bd001c4-f32d-4b82-93fc-b5d7ed46c031) | ![Screenshot 2023-08-28 at 9 51 41 AM](https://github.com/linode/manager/assets/114685994/5d3b8590-83de-493a-b20f-2cefb4b6711f) |
| Backups Drawer | ![Screenshot 2023-08-28 at 11 44 49 AM](https://github.com/linode/manager/assets/114685994/b6718742-93cb-4a44-acd3-4a6afd4e0537) | ![Screenshot 2023-08-28 at 11 08 20 AM](https://github.com/linode/manager/assets/114685994/0acde28f-1000-4191-b920-0f23c0ab4206) |


## How to test 🧪
1. **How to setup test environment?**
- Ensure the `dcSpecificPricing` feature flag is toggled on.
- Turn the MSW on.

2. **How to verify changes?**

- Make the changes described below to the mocks.
- Test that this mocked linode in Jakarta has DC-specific backups pricing (should use mocked `region_price`) in all the locations mentioned in the table above.
- Toggle the `dcSpecificPricing` flag off and confirm the mocked linode in Jakarta does _not_ have any increases in its backup price (should use mocked base price).
- Remove the region you added earlier from the mocked request`rest.get('*/linode/instances/:id'` and confirm that a linode not in Jakarta or Sao Paulo does not have any increases in its backup price (should use mocked base price).
- Ensure no new regressions in unit and e2e tests.

Update the MSW mocked request`rest.get('*/linode/instances/:id'` to include `region: 'id-cgk'` in the mocked linode created with the factory.

To test Backups Drawer:
Recommend modifying `rest.get('*/linode/instances'` to use a shorter list of linodes:
```
      linodeFactory.build({
        backups: { enabled: false },
        label: 'eu-linode',
        region: 'eu-west',
      }),
      linodeFactory.build({
        backups: { enabled: false },
        label: 'DC-Specific Pricing Linode',
        region: 'id-cgk',
      }),
```

3. **How to run Unit or E2E tests?**
```
yarn test BackupLinodeRow utils AddOnsPanel linodes
```
